test_conformance/extensions/cl_khr_external_semaphore/test_external_semaphore_sync_fd.cpp

```diff
@@ -0,0 +1,122 @@
+//
+// Copyright (c) 2024 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "harness/typeWrappers.h"
+#include "harness/extensionHelpers.h"
+#include "harness/errorHelpers.h"
+
+// Test it is possible to export a semaphore to a sync fd and import the same
+// sync fd to a new semaphore
+int test_external_semaphores_import_export_fd(cl_device_id deviceID,
+                                              cl_context context,
+                                              cl_command_queue defaultQueue,
+                                              int num_elements)
+{
+    cl_int err;
+
+    if (!is_extension_available(deviceID, "cl_khr_external_semaphore"))
+    {
+        log_info(
+            "cl_khr_external_semaphore is not supported on this platoform. "
+            "Skipping test.\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    if (!is_extension_available(deviceID, "cl_khr_external_semaphore_sync_fd"))
+    {
+        log_info("cl_khr_external_semaphore_sync_fd is not supported on this "
+                 "platoform. Skipping test.\n");
+        return TEST_SKIPPED_ITSELF;
+    }
+
+    // Obtain pointers to semaphore's API
+    GET_PFN(deviceID, clCreateSemaphoreWithPropertiesKHR);
+    GET_PFN(deviceID, clEnqueueSignalSemaphoresKHR);
+    GET_PFN(deviceID, clEnqueueWaitSemaphoresKHR);
+    GET_PFN(deviceID, clGetSemaphoreHandleForTypeKHR);
+    GET_PFN(deviceID, clReleaseSemaphoreKHR);
+
+    // Create ooo queue
+    clCommandQueueWrapper queue = clCreateCommandQueue(
+        context, deviceID, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE, &err);
+    test_error(err, "Could not create command queue");
```

Comment:
I realize this came over from the previous test also, but do we ever check that out-of-order queues are supported?

Does this need to be an out-of-order queue, or would an in-order queue suffice, such as the one provided by the harness?

Reply:
Please correct me if I am wrong but I can't spot any contraindications for `CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE` to be used along with external semaphore. Spec only mentions "valid command-queue". On the other hand there are in the test source code two events signal/wait which indicates synchronization verification.

I wasn't able to find anything more from original PR #1428. I am leaving as is for now.

Reply:
It's not so much that out-of-order queues cannot work with external semaphores, it's more that this is an optional feature that may not be supported by all devices supporting external semaphores.  So, if we want to use out-of-order queues, at the very least we'll need to check for out-of-order queue support, and skip the test if out-of-order queues aren't supported.

Note, because the queue is out-of-order, and there are no event dependencies between the signal and the wait in the out-of-order queue, I think (in theory, at least) the wait could execute before the signal.  I haven't fully convinced myself, but it seems like this could be a problem.  If the queue were an in-order queue, we would instead guarantee that the signal occurs before the wait, which I believe is the intent.

Reply:
Yep, I think I see your point. `wait_event` is not really waiting for anything so in theory it could execute before the signal. I just pushed verification for `CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE` property support, let me know if I should replace out-of-order queue with regular one.

Reply:
I don't think out of order queues pose an issue, the spec says clEnqueueSignalSemaphore will not return until the semaphore is safe to wait.  

I think in-order queues would be better, despite not being a problem.  The test is testing for importing/exporting the semaphore, OOO queues are not required to test this, no reason to not run this test if an implementation only supports in order. 

The value of external semaphores is not queue synchronization, it is sync with other APIs, or perhaps other OpenCL contexts.

Reply:
I still think it would be better to use an in-order queue instead, so the test can run on more devices, but all of the test logic is correct now so I'm fine with this going in as-is.